src/Random-Core/Bag.extension.st

```diff
@@ -0,0 +1,18 @@
+Extension { #name : #Bag }
+
+{ #category : #'*Random-Core' }
+Bag >> atRandom: aGenerator [
+	"Answer a random element of the receiver. Uses aGenerator which
+    should be kept by the user in a variable and used every time. Use
+    this instead of #atRandom for better uniformity of random numbers because 
+	 only you use the generator. Causes an error if self has no elements."
+	
+	| rand index |
+	self emptyCheck.
+	rand := aGenerator nextInt: self size.
+	"overwritten to use a faster enumeration"
+	index := 0.
+	self doWithOccurrences: [ :key :count | 
+		index := index + count.
+		rand <= index ifTrue: [ ^ key ] ]
```

Comment:
could be done without `index`:

    (rand := rand - count) <= 0 ifTrue: [ ^key ]

Reply:
Very good @nicolas-cellier-aka-nice . I might suggest using `index` instead of `rand` for the variable. It hurts my eyes to reassign `rand`.

Reply:
I will first add the original suggestion as a new PR so we get it in Pharo9, them I open a new issue as the #atRandom:  in Collection should be improved the same.

Reply:
I did a PR: https://github.com/pharo-project/pharo/pull/5663